### PR TITLE
[Dropzone] Enable multiple file uploads

### LINKED ISSUE #7: Dropzone: handle multiple files
It would be nice if Dropzone could handle multiple files.

I was expecting it would as its parent type is the FileType, that allows setting `multiple` option to `true`.

### PR DESCRIPTION
**Pull request by daFish**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes 
| Tickets       | Fix #7  
| License       | MIT

This change adds the posibility to upload multiple files using the Dropzone component. This is based on the work by @yassinehamouten.

- [x] Tests
- [x] Make it configurable using the FormType

### CODE REVIEW

**Comment by daFish** [comment]
@tgalopin & @weaverryan Would you mind to review this PR? 😄

File: src/Dropzone/Resources/views/form_theme.html.twig [hunk 1]
```diff
@@ -2,6 +2,7 @@
     {%- set dataController = (attr['data-controller']|default('') ~ ' symfony--ux-dropzone--dropzone')|trim -%}
     {%- set attr = attr|merge({ 'data-controller': '', class: (attr.class|default('') ~ ' dropzone-input')|trim}) -%}
 
+
```
Comment (weaverryan) [inline src/Dropzone/Resources/views/form_theme.html.twig]: minor, but we can revert this change

File: src/Dropzone/Resources/assets/src/controller.ts [hunk 2]
```diff
@@ -42,31 +42,34 @@ export default class extends Controller {
...
+
+            // If the file is an image, load it and display it as preview
+            this.previewImageTarget.style.display = 'none';
+            if (file.type && file.type.indexOf('image') !== -1) {
+                this._populateImagePreview(file);
+            }
         }
 
-        this._dispatchEvent('dropzone:change', file);
+        this._dispatchEvent('dropzone:change', event.target.files);
```
Comment (weaverryan) [inline src/Dropzone/Resources/assets/src/controller.ts]: Let's not this in the CHANGELOG - that's a minor BC break.

File: src/Dropzone/Resources/assets/src/controller.ts [hunk 3]
```diff
@@ -42,31 +42,34 @@ export default class extends Controller {
         this.previewImageTarget.style.display = 'none';
         this.previewImageTarget.style.backgroundImage = 'none';
         this.previewFilenameTarget.textContent = '';
+        document.querySelectorAll('.dropzone-preview-image-container').forEach((e) => e.remove());
```
Comment (weaverryan) [inline src/Dropzone/Resources/assets/src/controller.ts]: Sorry if I'm missing it, where/what is creating this `.dropzone-preview-image-container` element?

File: src/Dropzone/Resources/assets/src/controller.ts [hunk 4]
```diff
@@ -42,31 +42,34 @@ export default class extends Controller {
...
+                return;
+            }
+
+            // Hide the input and placeholder
+            this.inputTarget.style.display = 'none';
+            this.placeholderTarget.style.display = 'none';
+
+            // Show the filename in preview
+            this.previewFilenameTarget.textContent = file.name;
+            this.previewTarget.style.display = 'flex';
```
Comment (weaverryan) [inline src/Dropzone/Resources/assets/src/controller.ts]: I haven't tried the code yet, but how does this work with multiple files? It looks like if I attach multiple files, only the last one's filename would show up (it would override the earlier ones). Or very possibly I am missing a detail :)

**Review by weaverryan — COMMENTED** [review]
Thanks for this! I have some questions - either we're not (yet) quite supporting the "preview" for multiple files, or I'm missing it! 

Cheers :)

**Comment by daFish** [comment]
After a small conversation with @weaverryan on Slack I plan to add the following options to this component:
* ability to specify the number of files
* enable adding additional files after files were already dropped

File: src/Dropzone/assets/src/controller.ts [hunk 5]
```diff
@@ -21,6 +21,12 @@ export default class extends Controller {
 
     static targets = ['input', 'placeholder', 'preview', 'previewClearButton', 'previewFilename', 'previewImage'];
 
+    static values = {
+        numberOfFiles: Number
+    }
+
+    readonly numberOfFilesValue: number;
```
Comment (weaverryan) [inline src/Dropzone/assets/src/controller.ts]: It doesn't look like this is used?

File: src/Dropzone/templates/form_theme.html.twig [hunk 6]
```diff
@@ -2,6 +2,7 @@
     {%- set dataController = (attr['data-controller']|default('') ~ ' symfony--ux-dropzone--dropzone')|trim -%}
     {%- set attr = attr|merge({ 'data-controller': '', class: (attr.class|default('') ~ ' dropzone-input')|trim}) -%}
 
+
```
Comment (weaverryan) [inline src/Dropzone/templates/form_theme.html.twig]: This can be removed now

File: src/Dropzone/src/Form/DropzoneType.php [hunk 7]
```diff
@@ -28,6 +28,7 @@ public function configureOptions(OptionsResolver $resolver)
             'attr' => [
                 'placeholder' => 'Drag and drop or browse',
             ],
+            'multiple' => false,
```
Comment (weaverryan) [inline src/Dropzone/src/Form/DropzoneType.php]: Is that needed? It's already the default of the parent type

**Review by weaverryan — COMMENTED** [review]
Hi!

I played with this locally, but it doesn't look like multiple previews show up - just the one. Is that my mistake? Or do we still need a good deal more work - like to have potentially multiple "preview" areas instead of just one?

Thanks!

File: src/Dropzone/assets/src/controller.ts (hunk 5 above)
Comment (daFish) [reply src/Dropzone/assets/src/controller.ts]: Already removed.

File: src/Dropzone/templates/form_theme.html.twig (hunk 6 above)
Comment (daFish) [reply src/Dropzone/templates/form_theme.html.twig]: Done.

File: src/Dropzone/src/Form/DropzoneType.php (hunk 7 above)
Comment (daFish) [reply src/Dropzone/src/Form/DropzoneType.php]: Has been removed.

**Comment by kristjan-kure** [comment]
Hi! We need multiple file upload. Who can help with this? I understand [daFish](https://github.com/daFish) has been developing it, but we need it live? Who on the Symfony development team can help us to work with this?

**Comment by daFish** [comment]
After some thoughts I'm not going to finish this. It is based upon the work of @yassinehamouten and I'd rather propose to go with uppy.io.

**Comment by simonsolutions** [comment]
sorry to pick up on the old PR. But the above mentioned missing preview on more than one file seems still not to be working. I tried the current version of UX Dropzone, the multiple in the form element works. It accepts multiple files and posts them along with the form. But the preview of more than one file and the capability of adding more files seems also be missing.

Is there any trick i am missing?
